### PR TITLE
Interstate edge validation issue

### PR DESCRIPTION
**Pull request by alexnick83**
While validating interstate edges, there can be issues with assignments of Scalar values to symbols. These issues arise due to:
- The equality method `__eq__(self, other)` of the typeclass class, when `other` is a Scalar (or other Data object).
- The method `result_type_of(lhs, *rhs)`, when lhs/rhs are not typeclass objects or symbols.

To resolve the first issue, we add a `ctype` property to the Data class, that redirects to `Data.dtype.ctype`. For the second issue, we check if lhs/rhs are data objects. If this is true, we substitute them, for the rest of the method, with their dtype.

### CODE REVIEW

File: dace/dtypes.py [hunk 1]
```diff
@@ -403,9 +403,10 @@ def result_type_of(lhs, *rhs):
 
     rhs = rhs[0]
 
-    # Extract the type if symbolic
-    lhs = lhs.dtype if type(lhs).__name__ == 'symbol' else lhs
```
Comment (alexnick83) [inline dace/dtypes.py:407]: Why do we check the name of the type, instead of something along the lines of `isinstance(lhs, (dace.symbol, sympy.symbol))`?

Comment (tbennun) [reply dace/dtypes.py:408]: Import loop

Comment (tbennun) [reply dace/dtypes.py:408]: Import loop

File: dace/data.py [hunk 2]
```diff
@@ -126,6 +126,14 @@ def __repr__(self):
     def veclen(self):
         return self.dtype.veclen if hasattr(self.dtype, "veclen") else 1
 
+    # @property
+    # def type(self):
+    #     return self.dtype.type
```
Comment (tbennun) [inline dace/data.py:131]:
```suggestion
```

**Review by tbennun — CHANGES_REQUESTED** [review]
super minor change

**Review by tbennun — APPROVED** [review]

**Pull request merged**